vault/seal/ocikms/ocikms.go

```diff
@@ -0,0 +1,414 @@
+
+	cryptoEndpoint     string // OCI KMS crypto endpoint
+	managementEndpoint string // OCI KMS management endpoint
+
+	cryptoClient     *keymanagement.KmsCryptoClient     // OCI KMS crypto client
+	managementClient *keymanagement.KmsManagementClient // OCI KMS management client
+
+	currentKeyID *atomic.Value // Current key version which is used for encryption/decryption
+
+	logger log.Logger
+}
+
+var _ seal.Access = (*OCIKMSSeal)(nil)
+
+// NewSeal creates a new OCIKMSSeal seal with the provided logger
+func NewSeal(logger log.Logger) *OCIKMSSeal {
+	k := &OCIKMSSeal{
+		logger:       logger,
+		currentKeyID: new(atomic.Value),
+	}
+	k.currentKeyID.Store("")
+	return k
+}
+
+func (k *OCIKMSSeal) SetConfig(config map[string]string) (map[string]string, error) {
+	defer metrics.MeasureSince(metricInit, time.Now())
+	if config == nil {
+		config = map[string]string{}
+	}
+
+	// Check and set KeyID
+	switch {
+	case os.Getenv(EnvOCIKMSSealKeyID) != "":
+		k.keyID = os.Getenv(EnvOCIKMSSealKeyID)
+	case config[KMSConfigKeyID] != "":
+		k.keyID = config[KMSConfigKeyID]
+	default:
+		metrics.IncrCounter(metricInitFailed, 1)
+		return nil, fmt.Errorf("'%s' not found for OCI KMS seal configuration", KMSConfigKeyID)
+	}
+	k.logger.Info("OCI KMS configuration", KMSConfigKeyID, k.keyID)
+
+	// Check and set cryptoEndpoint
+	switch {
+	case os.Getenv(EnvOCIKMSCryptoEndpoint) != "":
+		k.cryptoEndpoint = os.Getenv(EnvOCIKMSCryptoEndpoint)
+	case config[KMSConfigCryptoEndpoint] != "":
+		k.cryptoEndpoint = config[KMSConfigCryptoEndpoint]
+	default:
+		metrics.IncrCounter(metricInitFailed, 1)
+		return nil, fmt.Errorf("'%s' not found for OCI KMS seal configuration", KMSConfigCryptoEndpoint)
+	}
+	k.logger.Info("OCI KMS configuration", KMSConfigCryptoEndpoint, k.cryptoEndpoint)
+
+	// Check and set managementEndpoint
+	switch {
+	case os.Getenv(EnvOCIKMSManagementEndpoint) != "":
+		k.managementEndpoint = os.Getenv(EnvOCIKMSManagementEndpoint)
+	case config[KMSConfigManagementEndpoint] != "":
+		k.managementEndpoint = config[KMSConfigManagementEndpoint]
+	default:
+		metrics.IncrCounter(metricInitFailed, 1)
+		return nil, fmt.Errorf("'%s' not found for OCI KMS seal configuration", KMSConfigManagementEndpoint)
+	}
+	k.logger.Info("OCI KMS configuration", KMSConfigManagementEndpoint, k.managementEndpoint)
+
+	// Check and set authTypeAPIKey
+	var err error
+	k.authTypeAPIKey = false
+	authTypeAPIKeyStr := config[KMSConfigAuthTypeAPIKey]
+	if authTypeAPIKeyStr != "" {
+		k.authTypeAPIKey, err = strconv.ParseBool(authTypeAPIKeyStr)
+		if err != nil {
+			metrics.IncrCounter(metricInitFailed, 1)
+			return nil, errwrap.Wrapf("failed parsing authTypeAPIKey parameter: {{err}}", err)
+		}
+	}
+	if k.authTypeAPIKey {
+		k.logger.Info("using OCI KMS with user principal")
+	} else {
+		k.logger.Info("using OCI KMS with instance principal")
+	}
+
+	// Check and set OCI KMS crypto client
+	if k.cryptoClient == nil {
+		kmsCryptoClient, err := k.getOCIKMSCryptoClient()
```

Comment:
There is a lot of duplicated logic with the encrypt method.  I think it would be much easier to just run a simple encrypt operation instead of duplicating the logic here.

Reply:
Could you please elaborate? What logic really duplicate here?